frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,342 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ReactElement, useCallback, useEffect, useState } from "react"
+
+import { useTheme } from "@emotion/react"
+import WaveSurfer from "wavesurfer.js"
+import RecordPlugin from "wavesurfer.js/dist/plugins/record"
+import { Delete } from "@emotion-icons/material-outlined"
+
+import { FileUploadClient } from "@streamlit/lib/src/FileUploadClient"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+import { AudioInput as AudioInputProto } from "@streamlit/lib/src/proto"
+import Toolbar, {
+  ToolbarAction,
+} from "@streamlit/lib/src/components/shared/Toolbar"
+import {
+  isNullOrUndefined,
+  labelVisibilityProtoValueToEnum,
+  notNullOrUndefined,
+} from "@streamlit/lib/src/util/utils"
+import { WidgetLabel } from "@streamlit/lib/src/components/widgets/BaseWidget"
+import { blend } from "@streamlit/lib/src/theme/utils"
+
+import { uploadFiles } from "./uploadFiles"
+import {
+  StyledAudioInputContainerDiv,
+  StyledWaveformContainerDiv,
+  StyledWaveformInnerDiv,
+  StyledWaveformTimeCode,
+  StyledWaveSurferDiv,
+} from "./styled-components"
+import NoMicPermissions from "./NoMicPermissions"
+import Placeholder from "./Placeholder"
+import {
+  BAR_GAP,
+  BAR_RADIUS,
+  BAR_WIDTH,
+  CURSOR_WIDTH,
+  STARTING_TIME_STRING,
+  WAVEFORM_PADDING,
+} from "./constants"
+import formatTime from "./formatTime"
+import AudioInputActionButton from "./AudioInputActionButtons"
+
+interface Props {
+  element: AudioInputProto
+  uploadClient: FileUploadClient
+  widgetMgr: WidgetStateManager
+  fragmentId?: string
+}
+
+const AudioInput: React.FC<Props> = ({
```

Comment:
Oh, I think one case that needs to be handled is the `disabled` state via the `disabled` prop that gets passed in from `widgetProps` in `ElementNodeRenderer`. I think you can just get it from the props here and use it instead of `element.disabled` everywhere.

Context: there are some niche situations where the frontend disables all widgets independent of what is configured by the user. This happens in a situation like when the app loses the connection.

Reply:
good catch!